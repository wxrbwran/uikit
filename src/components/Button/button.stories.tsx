import React from "react";
import { action } from "@storybook/addon-actions";

import Button from "./button";
import "../../styles/index.scss";

export default {
  title: "Button",
  component: Button,
  parameters: {}
};

export const DefaultButton = () => (
  <Button onClick={action("clicked")}>Hello Button</Button>
);

export const ButtonWithSize = () => (
  <>
    <Button onClick={action("clicked")} size="sm">
      Small
    </Button>
    <Button onClick={action("clicked")} size="lg">
      Large
    </Button>
  </>
);

export const ButtonWithType = () => (
  <>
    <Button onClick={action("clicked")}>Default</Button>
    <Button onClick={action("clicked")} btnType="primary">
      Primary
    </Button>
    <Button onClick={action("clicked")} btnType="danger">
      Danger
    </Button>
    <Button
      onClick={action("clicked")}
      btnType="link"
      target="_blank"
      href="https://www.baidu.com"
    >
      Link
    </Button>
  </>
);
