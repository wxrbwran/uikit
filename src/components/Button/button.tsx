import React, { FC, ButtonHTMLAttributes, AnchorHTMLAttributes } from "react";
import classnames from "classnames";

export type ButtonSize = "lg" | "sm";

export type ButtonType = "primary" | "default" | "danger" | "link";

interface BaseButtonProps {
  className?: string;
  /**设置Button禁用 */
  disabled?: boolean;
  /**设置Button的大小 */
  size?: ButtonSize;
  /**设置Button的类型 */
  btnType?: ButtonType;
  children: React.ReactChild;
  href?: string;
}

type NativeButtonProps = BaseButtonProps & ButtonHTMLAttributes<HTMLElement>;
type AnchorButtonProps = BaseButtonProps & AnchorHTMLAttributes<HTMLElement>;

export type ButtonProps = Partial<NativeButtonProps> &
  Partial<AnchorButtonProps>;
/**
 * 这是第一个Button组件
 *
 *
 */
export const Button: FC<ButtonProps> = props => {
  const {
    btnType,
    disabled,
    size,
    children,
    className,
    href,
    ...restProps
  } = props;
  // const a = 2;
  // if (a == 2) {
  //   console.log(222);
  // }
  const classes = classnames("btn", className, {
    [`btn-${btnType}`]: btnType,
    [`btn-${size}`]: size,
    disabled: btnType === "link" && disabled
  });
  if (btnType === "link" && href) {
    return (
      <a className={classes} href={href} {...restProps}>
        {children}
      </a>
    );
  }
  return (
    <button className={classes} disabled={disabled} {...restProps}>
      {children}
    </button>
  );
};

Button.defaultProps = {
  disabled: false,
  btnType: "default"
};

export default Button;
