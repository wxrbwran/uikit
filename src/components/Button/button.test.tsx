import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Button, { ButtonProps } from "./button";

const defaultProps = {
  onClick: jest.fn(),
};

const testProps: ButtonProps = {
  btnType: "primary",
  size: "lg",
  className: "klass",
};

const disabledProps: ButtonProps = {
  disabled: true,
  onClick: jest.fn(),
};

describe("测试Button组件", () => {
  it("渲染一个default Button", () => {
    const wrapper = render(<Button {...defaultProps}>Nice</Button>);
    const element = wrapper.getByText("Nice") as HTMLButtonElement;
    // console.log(element);
    expect(element).toBeInTheDocument();
    expect(element.tagName).toEqual("BUTTON");
    expect(element).toHaveClass("btn btn-default");
    expect(element.disabled).toBeFalsy();
    fireEvent.click(element);
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
  });
  it("渲染 Button, 基于不同的props", () => {
    const wrapper = render(<Button {...testProps}>Nice</Button>);
    const element = wrapper.getByText("Nice");
    expect(element).toBeInTheDocument();
    expect(element).toHaveClass("btn btn-primary btn-lg klass");
  });
  it("渲染 Link Button", () => {
    const wrapper = render(
      <Button {...testProps} btnType="link" href="http://xiaoran.work">
        Link
      </Button>
    );
    const element = wrapper.getByText("Link");
    // console.log(element);
    expect(element).toBeInTheDocument();
    expect(element.tagName).toEqual("A");
    expect(element).toHaveClass("btn btn-link");
  });
  it("渲染 Disabled Button", () => {
    const wrapper = render(<Button {...disabledProps}>Nice</Button>);
    const element = wrapper.getByText("Nice") as HTMLButtonElement;
    expect(element).toBeInTheDocument();
    expect(element.disabled).toBeTruthy();
    fireEvent.click(element);
    expect(disabledProps.onClick).not.toHaveBeenCalled();
  });
});
