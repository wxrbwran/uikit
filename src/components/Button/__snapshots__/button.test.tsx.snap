// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试Button组件 渲染 Disabled Button 1`] = `
<button
  class="btn btn-default"
  disabled=""
>
  Nice
</button>
`;
